runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -0,0 +1,325 @@
+
+    private final Object physicalLock = new Object();
+    private volatile UpdateT committedPhysical;
+    private final ConcurrentMap<CommittedBundle<?>, UpdateT> uncommittedPhysical =
+        new ConcurrentHashMap<>();
+
+    public DirectMetric(UpdateT zero) {
+      committedLogical = zero;
+      committedPhysical = zero;
+    }
+
+    protected abstract UpdateT combine(Iterable<UpdateT> updates);
+    protected abstract ResultT extract(UpdateT data);
+
+    public void updatePhysical(CommittedBundle<?> bundle, UpdateT tentativeCumulative) {
+      uncommittedPhysical.put(bundle, tentativeCumulative);
+    }
+
+    public void commitPhysical(final CommittedBundle<?> bundle, final UpdateT finalCumulative) {
+      // To prevent a query from blocking the commit, we perform the commit in two steps.
+      // 1. We perform a non-blocking write to the uncommitted table to make the new vaule
+      //    available immediately.
+      // 2. We submit a runnable that will commit the update and remove the tentative value in
+      //    a synchronized block.
+      uncommittedPhysical.put(bundle, finalCumulative);
+      COUNTER_COMMITTER.submit(new Runnable() {
+        @Override
+        public void run() {
+          synchronized (physicalLock) {
+            committedPhysical = combine(Arrays.asList(committedPhysical, finalCumulative));
+            uncommittedPhysical.remove(bundle);
+          }
+        }
+      });
+    }
+
+    public ResultT extractPhysical() {
+      ArrayList<UpdateT> updates = new ArrayList<>(uncommittedPhysical.size() + 1);
+      // Within this block we know that will be consistent. Specifically, the only change that can
+      // happen concurrently is the addition of new (larger) values to uncommittedPhysical.
+      synchronized (physicalLock) {
+        updates.add(committedPhysical);
+        updates.addAll(uncommittedPhysical.values());
+      }
+      return extract(combine(updates));
+    }
+
+    public void updateLogical(CommittedBundle<?> bundle, UpdateT tentativeCumulative) {
+      uncommittedLogical.put(bundle, tentativeCumulative);
+    }
+
+    public void commitLogical(final CommittedBundle<?> bundle, final UpdateT finalCumulative) {
+      // To prevent a query from blocking the commit, we perform the commit in two steps.
+      // 1. We perform a non-blocking write to the uncommitted table to make the new vaule
+      //    available immediately.
+      // 2. We submit a runnable that will commit the update and remove the tentative value in
+      //    a synchronized block.
+      uncommittedLogical.put(bundle, finalCumulative);
+      COUNTER_COMMITTER.submit(new Runnable() {
+        @Override
+        public void run() {
+          synchronized (logicalLock) {
+            committedLogical = combine(Arrays.asList(committedLogical, finalCumulative));
+            uncommittedLogical.remove(bundle);
+          }
+        }
+      });
+    }
+
+    public ResultT extractLogical() {
```

Comment:
This returns the tentative logical value rather than the committed, which doesn't seem correct.

Additionally, we never call updateLogical, so this really should just be an atomic `compareAndSet`